scls-cbor/src/Cardano/SCLS/CBOR/Canonical/Encoder.hs

```diff
@@ -224,43 +233,75 @@ instance
 -- Lists
 --------------------------------------------------------------------------------
 
--- | We always encode lists with the indefinite length encoding.
+-- | We always encode lists with the definite length encoding.
 instance (ToCanonicalCBOR v a) => ToCanonicalCBOR v [a] where
   toCanonicalCBOR v xs =
-    E.encodeListLenIndef
-      <> foldr (\x r -> toCanonicalCBOR v x <> r) E.encodeBreak xs
+    (unsafeToCanonicalEncoding (E.encodeListLen (fromIntegral $ length xs))) <> foldMap (toCanonicalCBOR v) xs
 
 instance (ToCanonicalCBOR v a) => ToCanonicalCBOR v (Seq.Seq a) where
   toCanonicalCBOR v xs =
-    E.encodeListLenIndef
-      <> foldr (\x r -> toCanonicalCBOR v x <> r) E.encodeBreak xs
+    (unsafeToCanonicalEncoding (E.encodeListLen (fromIntegral $ length xs))) <> foldMap (toCanonicalCBOR v) xs
 
 --------------------------------------------------------------------------------
 -- Maps
 --------------------------------------------------------------------------------
 
--- | We always encode maps with the indefinite length encoding.
+-- | We always encode maps with the definite length encoding and ordered by encoded keys byte-order.
 instance
   (ToCanonicalCBOR v k, ToCanonicalCBOR v val) =>
   ToCanonicalCBOR v (Map.Map k val)
   where
   toCanonicalCBOR v m =
-    E.encodeMapLenIndef
-      <> foldr
-        (\(k, val) b -> E.encodePreEncoded k <> toCanonicalCBOR v val <> b)
-        E.encodeBreak
-        mSorted
+    encodeAsMap l
    where
-    -- Order map by the byte-wise ordering of the canonically encoded map keys
-    mSorted =
-      List.sortOn fst $
-        Map.foldlWithKey'
-          ( \acc k val ->
-              let keyBytes = toStrictByteString $ toCanonicalCBOR v k
-               in (keyBytes, val) : acc
-          )
-          []
-          m
+    l =
+      Map.foldlWithKey'
+        (\acc k val -> SomeEncodablePair v k val : acc)
+        []
+        m
+
+{- | An existential wrapper for a key-value pair where both the key and value
+can be canonically CBOR-encoded under the version @v@.
+
+This type is useful for encoding heterogeneous collections of key-value pairs
+as CBOR maps, where the key and value types may vary but all support
+'ToCanonicalCBOR' for the same version.
+
+Use 'mkEncodablePair' to construct values of this type.
+-}
+data SomeEncodablePair v where
+  SomeEncodablePair :: (ToCanonicalCBOR v k, ToCanonicalCBOR v val) => proxy v -> k -> val -> SomeEncodablePair v
```

Comment:
Do we really need to keep `proxy` here? Seems that all the required information is in type so we can always recreate a proxy at the use case site (and get it immediately eliminated even on the compilation phase)?